### DISCUSSION TITLE
BSH - Token Transfer Service

### DISCUSSION THREAD

**Discussion by simsonraj** [icon-project/btp#32] [Q&A]
1. We assume that BEP20 and ERC20 are identical, but good to clarify in the BTP Requirements document this BEP20 is what is being implemented for the BSH token transfer service in solidity.

2. As in case of tokenBSH, all the balances are rightly transferred from the BSH to the token holder during a successfull transfer/ handleBTPError, the lockedbalance is rightfully accounted at the end of the transfer flow.
There is a requirement in the Token BSH for the User to be able to withdraw tokens.  Which balance would that be?

**Comment by cow-hs** [comment]
1. I agree with your opinion.
We plan to connect IRC2 <-> BEP20 <-> ERC20.
We will reflect on your requirements.

2. Token BSH transfers the balance in TokenBSH to Token Balance through a reclaim action.

**Reply by simsonraj** [reply]
@leeheonseung could you please confirm which of the below scenario is the reclaim function doing. also if the understanding is correct.

Scenario 1:
1. ICON's Alice sends A tokens,  _amount =10 to BSC's BOB by depositing to ICON's T_BSH  using tokenFallBack method.
2.  The BSC's BSH's handleBTPMessage intercepts & adds the 10 A tokens to BOB's **BSH token balance** and still holds it in the BSH
3. During Reclaim action the 10 A tokens from BSH is transferred to BOB 

Scenario 2:
1. ICON's Alice sends A tokens, _amount =10 to BSC's BOB by depositing to ICON's T_BSH  using tokenFallBack method.
2.  The BSC's BSH's handleBTPMessage intercepts & transfers 10 A tokens to BOB's  **account balance**
3. During Reclaim action, amount from refundable balance(as a result of token transfer failure) is transferred back to the user's tokens contract.

**Reply by cow-hs** [reply]
There are still internal discussions regarding BSH.
Please note that there may be changes to the following contents.

Scenario 1 is correct.

In case of case Scenario 2, assets are issued and then send to Bob through T_BSH only when the BTP message is successful.
In case of failure, the operation is not performed, and Alice only needs to reclaim the locked asset.
